### PR TITLE
Make SqlFragmentExpression constructor public

### LINKED ISSUE #17918: SqlFragmentExpression needs public constructor
Current constructor is internal and customers need to use it for custom translation of DbFunction.

Further, it may be useful to create some abstraction or helper methods to facilitate writing custom translation.

### PR DESCRIPTION
**Pull request by roji**
Closes #17918

With more issues like this I can rack up my closed issue count in no time!

### CODE REVIEW

**Comment by smitpatel** [comment]
build failures.

**Comment by smitpatel** [comment]
Also, add a test for UDF function mapping where this is used so we don't break it again unintentionally.

File: test/EFCore.SqlServer.FunctionalTests/Query/UdfDbFunctionSqlServerTests.cs [hunk 1]
```diff
@@ -14,7 +14,7 @@ public UdfDbFunctionSqlServerTests(SqlServer fixture, ITestOutputHelper testOutp
             : base(fixture)
         {
             Fixture.TestSqlLoggerFactory.Clear();
-            //Fixture.TestSqlLoggerFactory.SetTestOutputHelper(testOutputHelper);
+            Fixture.TestSqlLoggerFactory.SetTestOutputHelper(testOutputHelper);
```
Comment (smitpatel) [inline test/EFCore.SqlServer.FunctionalTests/Query/UdfDbFunctionSqlServerTests.cs]: Comment!

Comment (roji) [reply test/EFCore.SqlServer.FunctionalTests/Query/UdfDbFunctionSqlServerTests.cs]: Yes yes...

**Comment by roji** [comment]
Build failure because of Cosmos (Status code: TooManyRequests :face_with_thermometer:), merging.

**Pull request merged**